### PR TITLE
[FLINK-5941] Integrate Archiver pattern into handlers

### PR DESCRIPTION
**Pull request by zentol**
This PR adds the last piece to the foundation of the HistoryServer. ~~It builds on top of #3365 and #3376, meaning that only 58649fb is relevant for this PR.~~

The key addition is the ```JsonArchivist``` interface. An ```JsonArchivist``` generates Json responses and their respective REST paths for a given ```AccessExecutionGraph```.

The goal here is to allow writing these responses to files contained in a directory structure resembling the REST API, without defining the structure in 2 different places. With the additions made here the writing of files can be done in a simple loop:
```
AccessExecutiongraph graph = ...
for (JsonArchivistarchiver : WebRuntimeMonitor.getArchivers()) {
	for (ArchivedJson archive : archiver.archive(graph)) {
		File f = createFile(basePath, archive.path);
		writeJson(f, archive.json);
	}
}
```

Instead of extending existing job-related handlers I've added a new class in each of them that implements the new interface. This makes testing easier and also allows for easier refactoring/relocation.

### CODE REVIEW

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/WebRuntimeMonitor.java [hunk 1]
```diff
@@ -424,6 +425,33 @@ protected void initChannel(SocketChannel ch) {
 		LOG.info("Web frontend listening at " + address + ':' + port);
 	}
 
+	public static Archiver[] getArchivers() {
```
Comment (zentol) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/WebRuntimeMonitor.java]: i wanted to do something like this originally. However, since the archiving is done in the ``´MemoryArchivist``` i would have to pass the ```WebRuntimeMonitor``` reference to it. This is problematic since you now add an order in which they have to created. This specifically caused issues with the Yarn/Mesos runner that create the JobManager before the WebRuntimeMonitor.

It's a stupid problem and i agree that the static method approach is rather brittle, but i couldn't find a good solution for the above issue.

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/CurrentJobsOverviewHandler.java [hunk 2]
```diff
@@ -121,6 +125,26 @@ public String handleJsonRequest(Map<String, String> pathParams, Map<String, Stri
 		}
 	}
 
+	public static class CurrentJobsOverviewArchiver implements Archiver {
+
+		@Override
+		public ArchivedJson[] archiveJsonWithPath(AccessExecutionGraph graph) throws IOException {
```
Comment (zentol) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/CurrentJobsOverviewHandler.java]: I don't quite understand, we are using it? The archiver follows the code path of the handler in case ```includeRunningJobs``` and ```includeFinishedJobs``` are both true, since this is the only path that the WebFrontend actually calls.

I don't remember what in the end cause the job status to not appear, it is fixed now though.

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/WebRuntimeMonitor.java (hunk 1 above)
Comment (uce) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/WebRuntimeMonitor.java]: I see... changing that will probably be an unrelated refactoring. So let's leave it as is and add a comment. I think it won't even hurt to add a note why it is a static method etc.

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/CurrentJobsOverviewHandler.java (hunk 2 above)
Comment (uce) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/CurrentJobsOverviewHandler.java]: OK sorry I was confused by the manual JSON creation.

**Comment by uce** [comment]
Thanks! +1 to merge

**Comment by zentol** [comment]
merging.
